---
title: "Calss17: Cloud SRA data analysis"
author: "Hanhee Jo, PID# A59017994"
format: pdf
toc: true
---

# Downstream Analysis

```{r}
folders <- list.files(pattern = "_quant")
files <- paste0(folders, "/abundance.h5")
```

```{r}
file.exists(files)
```
```{r}
names(files) <- sub("_quant","",folders)
files
```


Load up the tximport package
```{r}
library(tximport)

txi.kallisto <- tximport(files, type = "kallisto", txOut = TRUE)
```
# Remove zero count genes
Before subsequent analysis, we might want to filter out those annotated transcripts with no reads:

```{r}
to.keep <- rowSums(txi.kallisto$counts) > 0
kset.nonzero <- txi.kallisto$counts[to.keep,]
```

```{r}
nrow(kset.nonzero)
```

and thos with no change over the samples
```{r}
keep2 <- apply(kset.nonzero,1,sd)>0
x <- kset.nonzero[keep2,]
```

# PCA

```{r}
pca <- prcomp(t(x), scale=TRUE)
summary(pca)
```

```{r}
mycols <- c("blue","blue","red","red")

library(ggplot2)
library(ggrepel)
ggplot(pca$x) + 
  aes(PC1, PC2) +
  geom_point(col=mycols)+
  geom_label_repel(label = rownames(pca$x), col=mycols)
```

# DESeq

```{r}
library(DESeq2)

sampleTable <- data.frame(condition = factor(rep(c("control", "treatment"), each = 2)))
rownames(sampleTable) <- colnames(txi.kallisto$counts)
```

```{r}
sampleTable
```

```{r}
dds <- DESeqDataSetFromTximport(txi.kallisto, colData = sampleTable, design = ~condition)
```

```{r}
dds <- DESeq(dds)
res <- results(dds)
```

```{r}
head(res)
```

